Reduce heuristic page cache size

chrisvest opened the pull request:
When `dbms.pagecache.memory` is not specified, and our configuration contain no other configurations from which we can derive it, then we would previously attempt to allocate 75% of the memory that would be available on the system after subtracting the maximum expected heap size.
It has turned out that this is sometimes a wee bit aggressive, so in this change we are reducing the heuristic to 50% of the available memory after heap.

chrisvest committed:
Reduce heuristic page cache size

When `dbms.pagecache.memory` is not specified, and our configuration contain no other configurations from which we can derive it, then we would previously attempt to allocate 75% of the memory that would be available on the system after subtracting the maximum expected heap size.
It has turned out that this is sometimes a wee bit aggressive, so in this change we are reducing the heuristic to 50% of the available memory after heap.

chrisvest committed:
Also update the documentation and distributed properties files with the new page cache memory heuristic

chrisvest committed:
Fix failing tests after the minimum page cache size was increased

lutovich commented on community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseSettings.java:
```diff
@@ -316,9 +316,11 @@ private static String defaultPageCacheMemory()
-                    long heuristic = (long) ((physicalMemory - maxHeapMemory) * 0.75);
-                    long min = 32 * 1024 * 1024; // We'd like at least 32 MiBs.
-                    long max = 1024 * 1024 * 1024 * 1024L; // Don't heuristically take more than 1 TiB.
+                    double ratioOfFreeMem = 0.50;
```
Maybe make this a constant somewhere?

chrisvest replied:
The only other uses are in comments and documentation, though.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseSettings.java:
```diff
@@ -316,9 +316,11 @@ private static String defaultPageCacheMemory()
-                    long max = 1024 * 1024 * 1024 * 1024L; // Don't heuristically take more than 1 TiB.
+                    double ratioOfFreeMem = 0.50;
+                    long heuristic = (long) ((physicalMemory - maxHeapMemory) * ratioOfFreeMem);
+                    long MiB = 1024 * 1024;
```
We already have corresponding constants in `Format`, even if `Format` is not really expected it, lets still use those.

chrisvest replied:
What would be a good place? Settings?

MishaDemianenko replied:
nope, KB is Settings unrelated in a same way as format, there is no way setting can influence that :)

Personally would prefer to have those constants in FileUtils, but that can be also influence of commons-io.

chrisvest replied:
That we don't know where to place the constants is a sign that they aren't really constants, but units of measure. Unfortunately Java has no system for expressing units. I can put them in FileUtils, but that place is just as bad as any other. I can alternatively create an enum for them, analogous to j.u.c.TimeUnit.

MishaDemianenko replied:
do not really see a problem with a language expressiveness
 Anyway, do we need to convert from one unit to another? if no, then would think that enum is a bit overkill

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/pagecache/ConfiguringPageCacheFactory.java:
```diff
@@ -126,12 +126,13 @@ public int calculatePageSize( Config config, PageSwapperFactory swapperFactory )
 
     public void dumpConfiguration()
     {
+        int MiB = 1024 * 1024;
```
We already have corresponding constants in `Format`, even if `Format` is not really expected it, lets still use those.

chrisvest committed:
Constantify number literals

chrisvest committed:
Introduce ByteUnit as a standard means of converting between quantities of bytes

This also fixes integer overflow bugs in the heuristic computation of the default value for the `dbms.pagecache.memory` setting.

MishaDemianenko commented on community/io/src/main/java/org/neo4j/io/ByteUnitSystem.java:
```diff
@@ -0,0 +1,60 @@
+    SI( 1000 ),
+    EIC( 1024 );
+
+    private final long increment;
```
does not look like increment is a correct name for it, more looks like `base` or `basis` imho

chrisvest replied:
Wikipedia uses the term "multiplier" which I think sounds good.

MishaDemianenko replied:
agree

MishaDemianenko commented on community/io/src/main/java/org/neo4j/io/ByteUnitSystem.java:
```diff
@@ -0,0 +1,60 @@
+ */
+public enum ByteUnitSystem
+{
+    SI( 1000 ),
```
why do we need SI system at all?

MishaDemianenko replied:
or let me change a question, why do we need both of them instead of using just one?

chrisvest replied:
Good question. I was trying to accurately model the domain, but this is probably a case of over-engineering. It will be useful if we use these units to interpret byte-valued settings, and we want to accurately understand the unit of the configured value. Does this sound like sufficient justification, or should  remove it?

MishaDemianenko replied:
i would vote for removing it because it's add a lot of additional complexity and tons of possible mistakes that can come later on after usage of non appropriate constant from different system

MishaDemianenko commented on community/io/src/main/java/org/neo4j/io/ByteUnit.java:
```diff
@@ -0,0 +1,221 @@
+     */
+
+    Byte( 0, ByteUnitSystem.SI, "B" ),
+    KiloByte( 1, ByteUnitSystem.SI, "kB" ),
```
i saw that we have kB before as well, maybe it's a good time to align it with all other units and change it to `KB` instead?

chrisvest replied:
No, small k is the standard abbreviation for kilo.

MishaDemianenko replied:
ok if that causing confusions only for me than it's fine :)

MishaDemianenko commented on community/io/src/main/java/org/neo4j/io/ByteUnit.java:
```diff
@@ -0,0 +1,221 @@
+ * A ByteUnit is a unit for a quantity of bytes.
+ *
+ * The unit knows how to convert between other units in its class, so you for instance can turn a number of KiBs into
+ * an accurate quantity of bytes. Some loss can occur when quantities are converted between units, due to integer
```
since we always returning long values as result of transformations, can you please or point me to a place where we can have rounding error or remove part that describe them

MishaDemianenko commented:
@chrisvest couple of comments, otherwise really nice conversion of units and we will remove a lot of hardcoded constants!

chrisvest committed:
Remove SI ByteUnits

We don't need them, so they were complicating things.
It was basically over-engineering to add them in the first place.

chrisvest commented:
@MishaDemianenko Should be good now.

davidegrohmann commented:
@chrisvest needs rebase
